src/test/java/com/google/devtools/build/lib/actions/ActionCacheCheckerTest.java

```diff
@@ -385,12 +442,475 @@ public void testDeletedConstantMetadataOutputCausesReexecution() throws Exceptio
+    String content2 = "content2";
+    Action action =
+        new InjectOutputFileMetadataAction(
+            output, createRemoteFileMetadata(content1), createRemoteFileMetadata(content2));
+
+    // Not cached
+    runAction(action);
+    writeContentAsLatin1(output.getPath(), content2);
+    // Not cached since local file changed
+    runAction(action);
+
+    assertStatistics(
+        0,
+        new MissDetailsBuilder()
+            .set(MissReason.NOT_CACHED, 1)
+            .set(MissReason.DIFFERENT_FILES, 1)
+            .build());
+    ActionCache.Entry entry = cache.get(output.getExecPathString());
+    assertThat(entry).isNotNull();
+    assertThat(entry.getOutputFile(output)).isEqualTo(createRemoteFileMetadata(content2));
+  }
+
+  @Test
+  public void saveOutputMetadata_treeMetadata_remoteFileMetadataSaved() throws Exception {
+    cacheChecker = createActionCacheChecker(/* storeOutputMetadata= */ true);
+    SpecialArtifact output =
+        createTreeArtifactWithGeneratingAction(artifactRoot, PathFragment.create("bin/dummy"));
+    ImmutableMap<String, RemoteFileArtifactValue> children =
+        ImmutableMap.of(
+            "file1", createRemoteFileMetadata("content1"),
+            "file2", createRemoteFileMetadata("content2"));
+    Action action =
+        new InjectOutputTreeMetadataAction(
+            output, createTreeMetadata(output, children, Optional.empty()));
+
+    // Not cached.
+    runAction(action);
+
+    assertThat(output.getPath().exists()).isFalse();
+    ActionCache.Entry entry = cache.get(output.getExecPathString());
+    assertThat(entry).isNotNull();
+    assertThat(entry.getOutputTree(output))
+        .isEqualTo(
+            SerializableTreeArtifactValue.create(children, Optional.empty()));
+    assertStatistics(0, new MissDetailsBuilder().set(MissReason.NOT_CACHED, 1).build());
+  }
+
+  @Test
+  public void saveOutputMetadata_treeMetadata_remoteArchivedArtifactSaved() throws Exception {
+    cacheChecker = createActionCacheChecker(/* storeOutputMetadata= */ true);
+    SpecialArtifact output =
+        createTreeArtifactWithGeneratingAction(artifactRoot, PathFragment.create("bin/dummy"));
+    Action action =
+        new InjectOutputTreeMetadataAction(
+            output,
+            createTreeMetadata(
+                output, ImmutableMap.of(), Optional.of(createRemoteFileMetadata("content"))));
+
+    // Not cached.
+    runAction(action);
+
+    assertThat(output.getPath().exists()).isFalse();
+    ActionCache.Entry entry = cache.get(output.getExecPathString());
+    assertThat(entry).isNotNull();
+    assertThat(entry.getOutputTree(output))
+        .isEqualTo(
+            SerializableTreeArtifactValue.create(
+                ImmutableMap.of(), Optional.of(createRemoteFileMetadata("content"))));
+    assertStatistics(0, new MissDetailsBuilder().set(MissReason.NOT_CACHED, 1).build());
+  }
+
+  @Test
+  public void saveOutputMetadata_emptyTreeMetadata_notSaved() throws Exception {
```

Comment:
This is hardly an important case, but does this mean that empty trees will not get action cache hits? I would think that actions generating those should also be eligible.

Reply:
The tree metadata of that empty tree won't be saved but it will get action cache hits. Updated the checks.

Reply:
Thanks!